components/dialog/dialog-confirm.js

```diff
@@ -82,17 +82,7 @@ class DialogConfirm extends DialogMixin(LitElement) {
 	}
 
 	_focusInitial() {
-		const footer = this.shadowRoot.querySelector('.d2l-dialog-footer-slot');
-		const nodes = footer.assignedNodes();
-		for (let i = 0; i < nodes.length; i++) {
-			const node = nodes[i];
-			if (node.nodeType !== Node.ELEMENT_NODE) continue;
-			if (!node.hasAttribute('primary')) {
-				forceFocusVisible(node);
```

Comment:
Does this still result in a visible focus ring now that the browser's autofocus is used instead of our `forceFocusVisible`?

Reply:
It does once the user presses Tab or Shift - I agree there's a bit of a tradeoff in that respect, but it's between having a visible focus ring and potentially hindering visually-impaired users from using the dialog effectively

Reply:
Hmm, I see. This change also effectively breaks all existing confirmation dialogs and requires that they all switch over to applying `autofocus` on the primary button... plus all new usages will need to do that as well -- something that could be easily missed or applied inconsistently to the wrong button.

Plus as you mentioned, this doesn't fix the underlying NVDA problem with all non-confirmation dialogs.

Add to that, `autofocus` itself is a little bit of a mystery to me... like [the docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button) describe it as:
> This Boolean attribute specifies that the button should have input focus when the page loads. Only one element in a document can have this attribute.

Both "when the page loads" and "only one element in a document" parts worry me. Showing/hiding the dialog doesn't constitute a "page load" (there are no iframes here) and a page could easily have multiple dialogs... it's unclear how the browser will handle this.

All this leads me to... I'm not sure if this is the best approach. The [dialog design pattern](https://www.w3.org/TR/wai-aria-practices/#dialog_modal) states:
> In all circumstances, focus moves to an element contained in the dialog.

If NVDA occasionally isn't handling this properly, that's a 3rd party bug with them and we should raise it there -- not try to work around it in our stuff.

Reply:
That's interesting, I'm curious if autofocus is working because the browser considers the dialog's shadow DOM instance to be a separate page or if it's just because the element is being patched in.

As far as NVDA is concerned, the "incorrect" reading seems like expected behavior, i.e. the dialog opens -> it's focused with JS -> NVDA skips over the dialog text to read the focused element. It's more of a mystery to me how dialogs are read correctly at all within the current setup, though I imagine it's some kind of race condition gimmick.

I definitely see how this could be a maintainability nightmare though. I can try and see if there's a way to add autofocus to incoming slot children automatically.

Reply:
Do any of the [W3C examples](https://www.w3.org/TR/wai-aria-practices/examples/dialog-modal/dialog.html) work properly? Because if they don't... then yeah I'm sorry but this is an NVDA thing.

The reason this normally works properly is that when focus moves into the dialog, the AT is supposed to notice that it's now inside an ARIA dialog (or now the native `<dialog>` element) and inform the user of that. It would be the same as other roles like menus where when you open it you move focus to the first menu item.

Reply:
NVDA does seem to be reading the modal dialog example correctly..... after doing some digging the incorrect read seems to be a byproduct of [the RAF autosizing fix](https://github.com/BrightspaceUI/core/blob/master/components/dialog/dialog-mixin.js#L318-L326), which focuses the button a split second after the dialog opens.

Reply:
A timing issue like this seems a lot more probable to me. @dbatiste that changed was [introduced here](https://github.com/BrightspaceUI/core/commit/e606c636873314d4fac3e5f3df4b5cd72165bbd0) to address some sizing issues. Not sure if we could move the focus bits outside the of rAF?

Reply:
It's possible that focus may not get applied correctly since the dialog `opened` (which changes the dialog's `display` property) state may not be reflected in the DOM yet, but it's worth a try.